File: src/agent/tools/builtin/write.py
"""Write tool: create or overwrite files."""

from __future__ import annotations

from pathlib import Path
from typing import Any


class WriteTool:
    @property
    def name(self) -> str:
        return "write"

    @property
    def description(self) -> str:
        return "Write content to a file, creating parent directories as needed."

    @property
    def input_schema(self) -> dict[str, Any]:
        return {
            "type": "object",
            "properties": {
                "file_path": {
                    "type": "string",
                    "description": "Path to the file to write.",
                },
                "content": {
                    "type": "string",
                    "description": "The content to write.",
                },
            },
            "required": ["file_path", "content"],
        }

    async def execute(self, *, file_path: str, content: str, **_: Any) -> str:
        path = Path(file_path).expanduser()
        try:
            path.parent.mkdir(parents=True, exist_ok=True)
            path.write_text(content)
            return f"Wrote {len(content)} bytes to {file_path}"
        except PermissionError:
            return f"Permission denied: {file_path}"
        except OSError as e:
            return f"Error writing {file_path}: {e}"

File: src/tools/builtin/write.ts
/** Write tool: create or overwrite files. */

import fs from 'node:fs';
import path from 'node:path';

import { expandUser } from '../../util/fsutil.js';
import type { Tool } from '../base.js';

export class WriteTool implements Tool {
  get name(): string {
    return 'write';
  }

  get description(): string {
    return 'Write content to a file, creating parent directories as needed.';
  }

  get inputSchema(): Record<string, any> {
    return {
      type: 'object',
      properties: {
        file_path: {
          type: 'string',
          description: 'Path to the file to write.',
        },
        content: {
          type: 'string',
          description: 'The content to write.',
        },
      },
      required: ['file_path', 'content'],
    };
  }

  async execute(params: Record<string, any>): Promise<string> {
    const filePath: string = params.file_path;
    const content: string = params.content ?? '';
    const resolved = expandUser(filePath);

    try {
      fs.mkdirSync(path.dirname(resolved), { recursive: true });
      fs.writeFileSync(resolved, content);
      return `Wrote ${content.length} bytes to ${filePath}`;
    } catch (e: any) {
      if (e?.code === 'EACCES') return `Permission denied: ${filePath}`;
      return `Error writing ${filePath}: ${e?.message ?? e}`;
    }
  }
}
